### DISCUSSION TITLE
How to inject formdata (or files in general)

### DISCUSSION THREAD

**Discussion by tombrouwer** [fastify/fastify#4479] [Q&A] [answered]
Hi, thanks for the hard work!

I try to test Fastify (with fastify multipart plugin). When I build the application and do a request (with insomnia/postman) it works fine. But I want to test the application with the `inject` method. But I can not figure it out how that will work. Can anyone help me with this?

Server:
```
const server = Fastify({ logger: { level: "debug" } });
server.register(multipart, { attachFieldsToBody: true });
server.post("/", { schema: {} }, async (request: FastifyRequest, reply: FastifyReply) => {
      const data = await request.file();
      console.log("data", data);
});
```
I tried a lot for the test so this is one of my attempts:
Test:
```
const fileData = fs.readFileSync(resolve(__dirname, `data/test.webp`));
const formData = new FormData();
formData.append("username", "nouser");
const blobblub = new Blob([fileData], { type: "image/webp" });
formData.append("filefield", blobblub);

const response = await server.inject({
    url: "/",
    method: "post",
    payload: formData,
    headers: { "Content-Type": 'multipart/form-data;boundary="boundary"' },
  });
```

If I not add `;boundary="boundary"` to the header, it will give an error (Multipart: Boundary not found). But when I send it whith Insomnia it will accept the content-type header with the value `multipart/form-data;`.

Maybe someone with experience in testing with inject and files?

thanks in advance!

**Comment by leandroandrade** [comment]
Hi, `form-auto-content` can help you with it. Try something like this:

```javascript
const formAutoContent = require('form-auto-content');

const form = formAutoContent({
      firstname: 'Foo',
      lastname: 'Bar',
      file: fs.createReadStream(path.join(__dirname, 'assets', 'fastify.pdf'))
  });

  const res = await fastify.inject({
      method: 'POST',
      url: '/upload-single',
      ...form
  });
```

**Comment by metcoder95** [comment, marked as answer]
The behavior is correct and intended, as by the spec [RFC2046](https://tools.ietf.org/html/rfc2046), the `boundary` property is mandatory. It is an arbitrary value added upon dispatch of a given request and should be attached at the end of the `MIME` of the `multipart/form-data` content type header.

It indicates the _boundary or end_ of a Request's body. For instance, it is required to allow the server to understand where the stream's body ends.

Now, why it works in `insomnia`? It is because `insomnia` does this calculation and handling for you; meanwhile, `light-my-request` (the engine behind `Fastify#inject`) does not do this calculation nor handling of the `boundary` delimiter.

As @leandroandrade suggested, the usage of the given library handles all the logic required as `insomnia` does under-the-hood
